Shared/Supporting Files/Views/CategoryView.swift

```diff
@@ -24,20 +24,53 @@ struct CategoryView: View {
     /// The search query in the search bar.
     @Binding private(set) var query: String
     
+    /// The samples to display in the name section of the search list.
+    @State private var nameSearchResults: [Sample] = []
+    
+    /// The samples to display in the description section of the search list.
+    @State private var descriptionSearchResults: [Sample] = []
+    
+    /// The samples to display in the tags section of the search list.
+    @State private var tagsSearchResults: [Sample] = []
+    
     /// A Boolean value that indicates whether to present the about view.
     @State private var isAboutViewPresented = false
     
-    /// The samples to display in the search list. Searching adjusts this value.
-    private var displayedSamples: [Sample] {
-        searchSamples()
-    }
+    /// The names of the samples already found in a previous section.
+    @State var previousSearchResults: Set<String> = []
     
     var body: some View {
         Group {
             if !isSearching {
                 CategoryGridView(samples: samples)
             } else {
-                SampleListView(samples: displayedSamples)
+                // The search results list.
+                List {
+                    if !nameSearchResults.isEmpty {
+                        Section(header: Text("Name Results")) {
+                            SampleListView(samples: nameSearchResults, query: query)
+                        }
+                    }
+                    if !descriptionSearchResults.isEmpty {
+                        Section(header: Text("Description Results")) {
+                            SampleListView(samples: descriptionSearchResults, query: query)
+                        }
+                    }
+                    if !tagsSearchResults.isEmpty {
+                        Section(header: Text("Tag Results")) {
+                            SampleListView(samples: tagsSearchResults, query: query)
+                        }
+                    }
+                }
+                .onChange(of: query) { _ in
+                    previousSearchResults = []
+                    nameSearchResults = searchSamplesNames()
+                    descriptionSearchResults = searchSamplesDescriptions()
+                    tagsSearchResults = searchSamplesTags()
```

Comment:
I might be playing devil's advocate here, but here are a few thoughts popped up:

- Try to make the search methods "[pure function](https://en.wikipedia.org/wiki/Pure_function)" so they don't have side effects. Sth like

```swift
    func searchNames(in samples: [Sample], with query: String) -> [Sample] {
        samples.filter { sample in
            sample.name.localizedCaseInsensitiveContains(query)
        }
    }
```

- How about the other logics? It may be better to just move them to the same file of the view. This also allows us to make the `previousSearchResults` a local variable, no need to be a state.

```swift
    private func searchSamples() async {
        // The names of the samples already found in a previous section.
        var previousSearchResults: Set<String> = []
        
        nameSearchResults = searchNames(in: samples, with: query)
        previousSearchResults.formUnion(nameSearchResults.map(\.name))
        
        descriptionSearchResults = searchDescriptions(in: samples, with: query)
            .filter { !previousSearchResults.contains($0.name) }
        previousSearchResults.formUnion(descriptionSearchResults.map(\.name))
        
        tagsSearchResults = searchTags(in: samples, with: query)
            .filter { !previousSearchResults.contains($0.name) }
    }
```

- Here though, I do hesitate on whether or not marking the helper method `searchSamples` as `async`. It is currently completely synchronous, and even with all 200-ish samples included in a future version, the time complexity would only be around O(3 * 200 * n) where n denotes the average length of a sample name / description / [tags]. That is a very short amount of time given the iOS 15+ device's speed. Still, it may be considered a "time consuming" task depend on how we view it. Let's don't add `async` for now.

Reply:
According to [the documentation](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/concurrency/), the `async` keyword is used "to indicate that a function or method is asynchronous." An asynchronous function/method is defined as "a special kind of function or method that can be suspended while it’s partway through execution." Also, "inside an asynchronous method, the flow of execution is suspended only when you call another asynchronous method — suspension is never implicit or preemptive — which means every possible suspension point is marked with `await`." Therefore a function/method should not be marked `async` if it does not await anything (ie. does not contain an `await` keyword), regardless of how long the function/method takes to execute.

Reply:
Thanks for the comment Phil. If I want to wrap a long-running synchronous call in Swift, is the correct way to use an actor or go back to [Thread](https://developer.apple.com/documentation/foundation/thread)?